ext/nnstreamer/tensor_decoder/tensordec-directvideo.c

```diff
@@ -30,74 +30,216 @@
 void fini_dv (void) __attribute__ ((destructor));
 
 #define DECODER_DV_VIDEO_CAPS_STR \
-    GST_VIDEO_CAPS_MAKE ("{ RGB, BGRx, GRAY8 }") \
+    GST_VIDEO_CAPS_MAKE ("{ GRAY8, RGB, BGR, RGBx, BGRx, xRGB, xBGR, RGBA, BGRA, ARGB, ABGR }") \
     ", views = (int) 1, interlace-mode = (string) progressive"
 
+/**
+ * @brief The supported video formats
+ */
+typedef enum
+{
+  DIRECT_VIDEO_FORMAT_UNKNOWN = 0,
+
+  /* Single Channel, Default: GRAY8 */
+  DIRECT_VIDEO_FORMAT_GRAY8 = 1,
+
+  /* 3 Channels, Default: RGB */
+  DIRECT_VIDEO_FORMAT_RGB = 2,
+  DIRECT_VIDEO_FORMAT_BGR = 3,
+
+  /* 4 Channels, Default: BGRx */
+  DIRECT_VIDEO_FORMAT_RGBx = 4,
+  DIRECT_VIDEO_FORMAT_BGRx = 5,
+  DIRECT_VIDEO_FORMAT_xRGB = 6,
+  DIRECT_VIDEO_FORMAT_xBGR = 7,
+  DIRECT_VIDEO_FORMAT_RGBA = 8,
+  DIRECT_VIDEO_FORMAT_BGRA = 9,
+  DIRECT_VIDEO_FORMAT_ARGB = 10,
+  DIRECT_VIDEO_FORMAT_ABGR = 11,
+} direct_video_formats;
+
+/**
+ * @brief Data structure for direct video options.
+ */
+typedef struct
+{
+  /* From option1 */
+  direct_video_formats format;
+} direct_video_ops;
+
+/**
+ * @brief List of the formats of direct video
+ */
+static const char *dv_formats[] = {
+  [DIRECT_VIDEO_FORMAT_UNKNOWN] = "UNKNOWN",
+  [DIRECT_VIDEO_FORMAT_GRAY8] = "GRAY8",
+  [DIRECT_VIDEO_FORMAT_RGB] = "RGB",
+  [DIRECT_VIDEO_FORMAT_BGR] = "BGR",
+  [DIRECT_VIDEO_FORMAT_RGBx] = "RGBx",
+  [DIRECT_VIDEO_FORMAT_BGRx] = "BGRx",
+  [DIRECT_VIDEO_FORMAT_xRGB] = "xRGB",
+  [DIRECT_VIDEO_FORMAT_xBGR] = "xBGR",
+  [DIRECT_VIDEO_FORMAT_RGBA] = "RGBA",
+  [DIRECT_VIDEO_FORMAT_BGRA] = "BGRA",
+  [DIRECT_VIDEO_FORMAT_ARGB] = "ARGB",
+  [DIRECT_VIDEO_FORMAT_ABGR] = "ABGR",
+  NULL,
+};
+
 /** @brief tensordec-plugin's GstTensorDecoderDef callback */
 static int
 dv_init (void **pdata)
 {
-  *pdata = NULL;                /* We have no internal data */
+  direct_video_ops *ddata;
+  ddata = *pdata = g_try_new0 (direct_video_ops, 1);
+
+  if (ddata == NULL) {
+    GST_ERROR ("Failed to allocate memory for decoder subplugin.");
+    return FALSE;
+  }
+
+  ddata->format = DIRECT_VIDEO_FORMAT_UNKNOWN;
+
   return TRUE;
 }
 
 /** @brief tensordec-plugin's GstTensorDecoderDef callback */
 static void
 dv_exit (void **pdata)
 {
-  /* Nothing to do */
+  g_free (*pdata);
   return;
 }
 
 /** @brief tensordec-plugin's GstTensorDecoderDef callback */
 static int
 dv_setOption (void **pdata, int opNum, const char *param)
 {
-  /* We do not accept anything. */
+  direct_video_ops *ddata;
+
+  if (!pdata || !*pdata) {
+    GST_ERROR ("There is no plugin data.");
+    return FALSE;
+  }
+
+  if (NULL == param || *param == '\0') {
+    GST_ERROR ("Please set the valid value at option1");
+    return FALSE;
+  }
+
+  ddata = *pdata;
+
+  /* When the dimension[0] is 4, the video format will be decided by option1. */
+  switch (opNum) {
+    case 0:
+    {
+      int f = find_key_strv (dv_formats, param);
+
+      ddata->format = (f < 0) ? f : DIRECT_VIDEO_FORMAT_UNKNOWN;
```

Comment:
oh...my fault. this is invalid condition.

Reply:
fixed!